Develop

ValeriyaKot opened the pull request.

ValeriyaKot committed:
Create project pitch

ValeriyaKot committed:
Create app users

ValeriyaKot committed:
Add serializers, utils, urls

ValeriyaKot committed:
Add AuthViewSet

ValeriyaKot committed:
Add users in settings, add include in urls

ValeriyaKot committed:
Add constants, add profile in models

ValeriyaKot committed:
Add profile in admin site

ValeriyaKot committed:
Add ProfileViewSet, ProfileSerializer

ValeriyaKot committed:
Create courses app

ValeriyaKot committed:
Add behavior, permissions

ValeriyaKot committed:
Add serializers, urls

ValeriyaKot committed:
Add material, module views and serializers

ValeriyaKot committed:
Delete Questions views, serializers

slava-pasedko commented on pitch-app/pitch/apps/users/serializers.py:
```diff
@@ -0,0 +1,75 @@
+        password_validation.validate_password(value)
+        return value
+
+class EmptySerializer(serializers.Serializer):
```
What is that

slava-pasedko commented on pitch-app/pitch/apps/users/serializers.py:
```diff
@@ -0,0 +1,75 @@
+
+    class Meta:
+        model = User
+        fields = ('id', 'username', 'email', 'first_name', 'last_name', 'is_active', 'is_staff', 'auth_token', 'profile')
```
Split the line so it won't be so long

slava-pasedko commented on pitch-app/pitch/apps/users/serializers.py:
```diff
@@ -0,0 +1,75 @@
+        fields = ('id', 'username', 'email', 'password', 'first_name', 'last_name', 'profile')
+
+    def create(self, validated_data):
+        profile_data = validated_data.pop('profile')
```
How does a password look in the admin if we open a freshly created user?

I don't see that you're actually hashing the password, f.e.:

```
from django.contrib.auth.hashers import make_password

validated_data['password'] = make_password(validated_data.get('password'))
```

Or is it being hashed somewhere else?

slava-pasedko commented on pitch-app/pitch/apps/courses/models.py:
```diff
@@ -0,0 +1,41 @@
+
+class Course(Timestampable):
+    teacher = models.ForeignKey(Profile, on_delete=models.CASCADE)
+    students = models.ManyToManyField(Profile, related_name='students_of_course')
```
related name "course_students" would be better

slava-pasedko commented on pitch-app/pitch/apps/courses/models.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class Answer(models.Model):
+    answer = models.TextField()
```
We'd need to determine somehow that the one of the answers is a right answer, don't we?

slava-pasedko commented on pitch-app/pitch/apps/courses/models.py:
```diff
@@ -0,0 +1,41 @@
+
+class Answer(models.Model):
+    answer = models.TextField()
+    question = models.ForeignKey(Question, on_delete=models.DO_NOTHING)
```
Why do we have DO_NOTHING here? Why do we need an answer for a question that's been deleted?

ValeriyaKot committed:
Add enroll_course

slava-pasedko commented on pitch-app/pitch/apps/courses/models.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class StudentAnswer(models.Model):
+    profile = models.ForeignKey(Profile, on_delete=models.DO_NOTHING)
```
Same thing with DO_NOTHING. 

If we're going to add an ability of deleting profile, question or answer we would do it using some bool field like "deleted" but won't delete from database

slava-pasedko commented on pitch-app/pitch/apps/courses/permissions.py:
```diff
@@ -0,0 +1,11 @@
+    def has_object_permission(self, request, view, obj):
+        if request.method in SAFE_METHODS:
+            return True
+        profile = Profile.objects.get(user=request.user)
```
What if get method doesn't find a profile? The DoesNotExist exception will occur. We'd need to handle that

slava-pasedko commented on pitch-app/pitch/apps/courses/serializers.py:
```diff
@@ -0,0 +1,69 @@
+    #         models.Module.objects.create(course=course, **module)
+    #     return course
+    #
+    # def update(self, instance, validated_data):
```
delete comments from PR

ValeriyaKot committed:
Delete EnrollSerializer

slava-pasedko commented on pitch-app/pitch/apps/courses/views.py:
```diff
@@ -0,0 +1,50 @@
+    serializer_class = serializers.MaterialSerializer
+    permission_classes = [IsAuthenticated, IsTeacherOrReadOnly]
+
+
```
pep8; install flake8 and check the repo before pushing & commiting using it

slava-pasedko commented on pitch-app/pitch/apps/courses/views.py:
```diff
@@ -0,0 +1,50 @@
+    @action(methods=['POST', ], detail=False)
+    def add_course(self, request):
+        serializer = serializers.CourseSerializer(data=request.data)
+        teacher = Profile.objects.get(user=request.user)
```
rename variable to user since at this exact line we don't know whether it student or a teacher yet

ValeriyaKot committed:
Refactor code

ValeriyaKot committed:
Add try/except in permissions, perform_create in views

slava-pasedko commented on pitch-app/pitch/apps/courses/permissions.py:
```diff
@@ -6,6 +6,9 @@ class IsTeacherOrReadOnly(BasePermission):
+            if profile.role == 'teacher':
+                return obj
+        except Profile.DoesNotExist:
+            raise ValueError('Profile does not exist')
```
Why raise ValueError? It's a permission file, where we should return a boolean value which would indicate if request is processed further. 

For example, if request method in SAFE_METHODS, we return True in order to allow performing this request.

If we have a profile, we should check if profile role is teacher. 

But anyway, we have to return True or False, so we can write this like:
`profile = Profile.objects.get(user=request.user)
            return profile.role == 'teacher'`

This way, if user is not a teacher, method will return False and will prevent such requests from users with a role that differ from "teacher".

If case where we don't have a profile, we don't really know what type of user this is. It means that we should prevent such request.

ValeriyaKot replied:
так а где должен в итоге находиться проверка на учителя в safe_methods?

ValeriyaKot committed:
Add requirements.txt

ValeriyaKot committed:
Add has_permission

ValeriyaKot committed:
Add permissions_by_action

ValeriyaKot commented:
I added has permission, and permissions by actions in CourseDetailView

slava-pasedko commented on pitch-app/pitch/apps/courses/urls.py:
```diff
@@ -0,0 +1,15 @@
+
+router = routers.SimpleRouter()
+router.register('pitch/courses', views.CourseView, basename='courses')
+router.register('pitch/edit_courses', views.CourseEditView, basename='edit_courses')
```
It would be better if course and other model views had only one url per view, f.e.

pitch/courses ---> GET (list), POST (create)
pitch/courses/<:pk>  --- GET (retrieve), PUT & PATCH (update, partial_update), DELETE (destroy)

ValeriyaKot committed:
Refactor

ValeriyaKot committed:
Delete Test, Answer models, views

ValeriyaKot committed:
Create quizzes app

slava-pasedko commented on pitch-app/pitch/apps/courses/views.py:
```diff
@@ -0,0 +1,64 @@
+            if profile.role == 'teacher':
+                serializer.save(teacher=profile)
+                return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response('Course can add only teacher', status=status.HTTP_400_BAD_REQUEST)
```
What response would I get if I sent non-valid data? F.e. serializer_is_valid() returns False.

I still get a 400 Bad Request error which is good, but there's a message that says "Course can add only teacher".

Maybe I am a teacher, maybe not, but the cause of 400 error was bad data rather than profile role

ValeriyaKot committed:
Refactor add_course

ValeriyaKot committed:
Add create in ModuleSerializer

ValeriyaKot committed:
Add signals.py

Change ForeignKey to OneToOneField in users.models, refactor register in view, RegisterSerializer

ValeriyaKot committed:
Add StudentAnswer model

ValeriyaKot committed:
Add write_student_answers in views

slava-pasedko committed:
Merge pull request #2 from ValeriyaKot/feature/quiz

Create quizzes app
